### DISCUSSION TITLE
3D Ablation simulation with MFEM

### DISCUSSION THREAD

**Discussion by ghost** [mfem/mfem#3543] [Q&A]
Hello MFEM community ! 👋 

I am currently working on a problem of heat conduction in a solid driven by an external heat flux. I got this part to work very nicely even for a multimaterial solid and temperature dependent properties.

The next step I am aiming at is ablation, and this is why I am here asking your help ! 🙏 

Let's say the algorithm is the following straightforward conditional:

> if a cell reaches a temperature that is equal to or greater than the fusion temperature of the material, we want to consider this cell is instantaneously completely ablated and we want to remove it from the solid.

How would you guys go around deleting cells from an instance of `mfem.Mesh` on the fly during an unsteady heat conduction solver is running ? 

- Is there a `.delete` somewhere that would do this magically ? 🧙‍♂️ 
- Would is be wiser to suddenly give that cell a conductivity that is so high that it is like it's not there anymore ?
- Option 3 ?

I can post parts of or the entirety of the solver I have right now if it can help figuring out the details !

Thank you so much ! Looking forward to hearing your ideas !

**Comment by vladotomov** [comment]
Hi @tbridel,

When you delete an element, what happens to the boundary conditions at its faces? Are you using H1 or L2 FE space?

**Reply by ghost** [reply]
Hi @vladotomov !

Thank you for coming back to me !

When at the end of iteration `n` some cells get deleted, the outer skin of the solid changes indeed and at iteration `n+1` the boundary condition is applied to the new outer skin. We go from

```
|****|****|****|****|
|____|____|____|____|
````

to

```
|****|    |****|****|
|____|****|____|____|
```

if, say, the second cell gets ablated, and the `*` represent the boundary condition (the flux be coming from the top).

I am in an H1 FE space for now. Could you please explain why this could be an important factor, I am not sure I understand why 😬

**Reply by vladotomov** [reply]
@tbridel 
Deleting elements from a mesh is possible, see `miniapps/meshiing/trimmer.cpp`. But you need more than that, because you have existing FE spaces and solutions and deleting an element would require fixing their internal structures.

One option is to look at creating a `SubMesh` (see `mesh/submesh/`) with proper boundary attributes, and then use `SubMesh::Transfer` for the solutions. You will have a different FESpace after every update. The advantage is that you'll be using the existing MFEM interface, but I don't know if it will be too slow with dynamic changes every time step.

The best approach might be to introduce some custom marking of inactive cells and active face-boundaries. Then you pick which elements must be assembled, which DOFs must be eliminated from the linear solve, which DOFs are Dirichlet BC, which faces get weak BC integrals, etc. It requires some bookkeeping, but it will work eventually. With this approach you don't modify the mesh, but eliminate processing of inactive elements.

I asked about L2/H1 because the marking approach would be easier to do for L2 as the DOFs are always element-local.

**Reply by ghost** [reply]
Hi @vladotomov ,

Thank you very much for your answer - my apologies for the delay.

I have to look into the idea with the `SubMesh`, but from what you describe it sound good. As for the dynamic changes, it will not actually happen at every time step in the unsteady heat conduction equation. The actual process is more like the following:

1. Receive a heat flux vector from a third party library, as well as an exposure time,
2. March the unsteady conduction in time for the duration of the exposure using a constant heat flux,
3. Consider the updated wall temperature after exposure, and ablate cells that have melted, if any,
4. Send the indices of the melted cells back to the third party library,
5. etc...

So maybe the cost of `SubMesh::Transfer` can actually be absorbed in the cost of all the steps 4., 5., etc... back to 1. executed by the third party library, if i trigger the transfer and the 4., 5., etc... steps asynchronously.

I will be away until April 11th or so, but I was wondering if it would be interesting for MFEM to have an ablation example ? I could write a mwe - with your help - and open a PR if it's something that you may be looking for ?

**Reply by vladotomov** [reply]
Sounds interesting, we'd be happy to help and add an example for dynamic submeshes. We'd prefer to be self-contained without 3rd party libraries, i.e., steps 1 and 4 to be something simple, but this should be easy to construct.

In the meantime you can also check `/miniapps/multidomain/` for example `SubMesh` use.

@mfem/editors can we please add @tbridel to the organization?

**Reply by ghost** [reply]
@vladotomov Thank you. Steps 1 and 4 _et sqq._ can be excluded actually, the heat flux can be set to a constant for each "cycle" and the cycle can go to step 3 only before looping back.

Thanks for the support anyways, I'll pick this up when I come back mid-April !

**Comment by colmenaresj** [comment]
Hi @tbridel , this sounds like an interesting project. Have you figured out how to do it? I wonder if you would benefit from using something more like a surface-tracking method, where the cells are fixed but you track the interface of the solid using a level-set method and then the surface would move/shrink based on some speed due to ablation. I haven't worked on ablation before, but something like that is used for simulating evaporation of a droplet, for example.  

Is the cell deletion method a standard approach for this kind of problem?
